src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/Tasks/CompilerLocalReference.cs

```diff
@@ -1,347 +1,352 @@
+                }
+                else
+                {
+                    sw.WriteLine(LocalApplicationFile.Serialize());
+                }
+
+                if (LocalMarkupPages != null && LocalMarkupPages.Length > 0)
+                {
+                    for (int i = 0; i < LocalMarkupPages.Length; i++)
+                    {
+                        sw.WriteLine(LocalMarkupPages[i].Serialize( ));
+                    }
+                }
+
+                sw.Flush();
+                _taskFileService.WriteFile(memStream.ToArray(), _localCacheFile);
+
+                bSuccess = true;
+            }
+
+            return bSuccess;
+        }
+        
+        //
+        // Read the Local Reference cache file, load the cached information
+        // to the corresponding data fields in this class.
+        //
+        internal bool LoadCacheFile()
+        {
+            Debug.Assert(String.IsNullOrEmpty(_localCacheFile) != true, "_localCacheFile must not be empty.");
+
+            bool loadSuccess = false;
+
+            Stream stream = _taskFileService.GetContent(_localCacheFile);
+
+            // using Disposes the StreamReader when it ends.  Disposing the StreamReader 
+            // also closes the underlying MemoryStream.  Don't look for BOM at the beginning
+            // of the stream, since we don't add it when writing.  TaskFileService takes care
+            // of this.
+            using (StreamReader srCache = new StreamReader(stream, false))
+            {
+                // The first line must be for InternalTypeHelperFile.
+                // The second line is for Local Application Defintion file.
+                // For Library, the second line is an empty line.
+
+                InternalTypeHelperFile = srCache.ReadLine();
+
+                string lineText;
+
+                lineText = srCache.ReadLine();
+
+                LocalApplicationFile = LocalReferenceFile.Deserialize(lineText);
+
+                ArrayList alMarkupPages = new ArrayList();
+
+                while (srCache.EndOfStream != true)
+                {
+                    lineText = srCache.ReadLine();
+                    LocalReferenceFile lrf = LocalReferenceFile.Deserialize(lineText);
+
+                    if (lrf != null)
+                    {
+                        alMarkupPages.Add(lrf);
+                    }
+                }
+
+                if (alMarkupPages.Count > 0)
+                {
+                    // With SDK style projects, a file may be removed from the Page glob (via deleting from disk).
+                    // When incremental build is triggered after that, we can potentially try to
+                    // compile a file that is no longer on disk.  So ensure that we don't queue any files
+                    // that no longer exist.
+                    LocalMarkupPages = alMarkupPages.Cast<LocalReferenceFile>().Where(x => File.Exists(x.FilePath)).ToArray();

```

Comment:
whoa! does this touch every markup file on the disk for every incremental build? that can't be good for build-perf 😢 

Can we instead use a [`FileSystemWatcher`](https://docs.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher?view=netcore-3.0) and let the OS tell us when something is deleted? (i.e., don't poll - use interrupts?)

/cc @nguerrera, @dsplaisted can you think of a more efficient way to do this? i'm afraid the answer is going to be no, but figured i'd check in anyway.

Reply:
The watcher would only run during build, the delete is prior to the build, so that can't help us here.  

I think any perf issue would be a one-shot cost that is of similar disk cost to just doing 2 back to back incremental builds.  In the incremental build case, we're still asking MarkupCompilePass1 to touch xaml files and re-generate BAML.

Reply:
I suppose we can (if we have access to it) examine the Page blob for files that no longer exist in it and skip them.  I can look into plumbing that down into this part of PBT.

Reply:
Correction above, the Page change is what was invalidating the markup and touching all XAML.  I got confused between some binlogs.  In any case, I have a candidate fix I am messing with right now that should just rely on an intersection between the set of current pages passed into MarkupCompilePass1 vs. the cached values.

Reply:
Resolving as I've force pushed a change that just hits the current Page blob.